snorkel/analysis/utils.py

```diff
@@ -12,21 +12,73 @@ def set_seed(seed: int) -> None:
     torch.manual_seed(seed)
 
 
-def probs_to_preds(probs: np.ndarray) -> np.ndarray:
+def break_ties(
+    Y_prob: np.ndarray, tie_break_policy: Optional[str] = "random"
+) -> np.ndarray:
+    """Break ties among probabilistic labels according to given policy.
+
+    Policies to break ties include:
+    "abstain": return an abstain vote (0)
+    "true-random": randomly choose among the tied options
+    "random": randomly choose among tied option using deterministic hash
+
+    NOTE: if tie_break_policy="true-random", repeated runs may have slightly different results due to difference in broken ties
+
+    Parameters
+    ----------
+    Y_prob
+        An [n,k] array of probabilistic labels
+    tie_break_policy
+        Policy to break ties, by default 'random'
+
+    Returns
+    -------
+    np.ndarray
+        An [n,1] array of integer labels
+    """
+
+    n, k = Y_prob.shape
+    Y_pred = np.zeros(n)
+    diffs = np.abs(Y_prob - Y_prob.max(axis=1).reshape(-1, 1))
+
+    TOL = 1e-5
+    for i in range(n):
+        max_idxs = np.where(diffs[i, :] < TOL)[0]
+        if len(max_idxs) == 1:
+            Y_pred[i] = max_idxs[0] + 1
+        # Deal with "tie votes" according to the specified policy
+        elif tie_break_policy == "random":
+            Y_pred[i] = max_idxs[i % len(max_idxs)] + 1
```

Comment:
Clever! This is much lighter-weight than hashing the whole label array.

Reply:
Hmmmm is there any way to check if the original label matrix row here was all zeros? That's probably the only time we want to hash the index instead of the row of labels. For example, if two data points both had label outputs [0, 1, 2] and lf_2_acc == lf_3_acc, I'd want the predictions for them to be the same. If they're all zeros, then sure, hash the ID so all of the uncovered points don't have the same class.

Reply:
is it weird to always prefer a class over the other in this case? say in the binary case, we have `probs=[0.5, 0.5]` and it's due to the LF accuracies that voted on this datapoint being the same. Then we'd end up with all datapoint having class=1 (or class=2), which doesn't necessarily reflect what "random" would be

Reply:
Yeah, I think I like the pseudo-random distribution. The intuitive check I'm using is "if I have 100 times, they should be split approximately 50/50 by class if I'm doing random tie-breaking", just in a way that's deterministic (which this achieves).

Reply:
i.e., I like the way it currently is in this PR.

Reply:
ok then to at least get rid of the cyclical artifacts, let's do `hash(str(i)) % n_ties`